subprojects/dependency-management/src/main/java/org/gradle/api/internal/std/DefaultDependenciesAccessors.java

```diff
@@ -263,29 +247,150 @@ public ClassPath getClasses() {
         return classes;
     }
 
-    private static class DependencyAccessorCompilationOperation implements RunnableBuildOperation {
+    private abstract class AbstractAccessorUnitOfWork implements UnitOfWork {
+        private static final String OUT_SOURCES = "sources";
+        private static final String OUT_CLASSES = "classes";
+        public static final String DEPENDENCY_ACCESSORS_PREFIX = "dependency-accessors/";
+
+        @Override
+        public Identity identify(Map<String, ValueSnapshot> identityInputs, Map<String, CurrentFileCollectionFingerprint> identityFileInputs) {
+            return () -> {
+                Hasher hasher = Hashing.sha1().newHasher();
+                identityInputs.values().forEach(s -> s.appendToHasher(hasher));
+                return hasher.hash().toString();
+            };
+        }
+
+        @Override
+        public <T> T withWorkspace(String identity, WorkspaceAction<T> action) {
+            return workspace.withWorkspace(DEPENDENCY_ACCESSORS_PREFIX + identity, (workspace, history) -> action.executeInWorkspace(workspace));
+        }
+
+        @Override
+        public Optional<ExecutionHistoryStore> getHistory() {
+            return Optional.of(workspace.getHistory());
+        }
+
+        protected abstract List<ClassSource> getClassSources();
+
+        @Override
+        public WorkOutput execute(@Nullable InputChangesInternal inputChanges, InputChangesContext context) {
+            File workspace = context.getWorkspace();
+            File srcDir = new File(workspace, OUT_SOURCES);
+            File dstDir = new File(workspace, OUT_CLASSES);
+            List<ClassSource> sources = getClassSources();
+            compile(srcDir, dstDir, sources, classPath);
+            return new WorkOutput() {
+                @Override
+                public WorkResult getDidWork() {
+                    return WorkResult.DID_WORK;
+                }
+
+                @Override
+                public Object getOutput() {
+                    return loadRestoredOutput(workspace);
+                }
+            };
+        }
+
+        @Override
+        public Object loadRestoredOutput(File workspace) {
+            File srcDir = new File(workspace, OUT_SOURCES);
+            File dstDir = new File(workspace, OUT_CLASSES);
+            return new GeneratedAccessors(srcDir, dstDir);
+        }
+
+        @Override
+        public void visitOutputs(File workspace, OutputVisitor visitor) {
+            File srcDir = new File(workspace, OUT_SOURCES);
+            File dstDir = new File(workspace, OUT_CLASSES);
+            visitOutputDir(visitor, srcDir, OUT_SOURCES);
+            visitOutputDir(visitor, dstDir, OUT_CLASSES);
```

Comment:
`visitOutputDir()` could look up the directories, too.

Reply:
Sorry what do you mean?

Reply:
You resolve `srcDir` using `OUT_SOURCES` and then pass it to `visitOutputDir()` that also knows about `OUT_SOURCES`, so you could move the resolving of the dir to that method.

Reply:
Oh I see, yes, thanks!